index.js

```diff
@@ -2120,24 +2124,41 @@
   //. Cons('red', Cons('blue', Cons('green', Nil)))
   //. ```
   function sortBy(f, foldable) {
-    var rs = reduce(function(xs, x) {
-      var fx = f(x);
-      var lower = 0;
-      var upper = xs.length;
-      while (lower < upper) {
-        var idx = Math.floor((lower + upper) / 2);
-        if (lte(xs[idx].fx, fx)) lower = idx + 1; else upper = idx;
-      }
-      xs.splice(lower, 0, {x: x, fx: fx});
+    var foldableIsArray = Array.isArray(foldable);
+    var idx = 0;
+    function reducer(xs, x) {
+      xs.push({idx: idx, x: x, prop: f(x)});
+      idx += 1;
       return xs;
-    }, [], foldable);
-
-    var F = foldable.constructor;
-    var result = empty(F);
-    for (var idx = 0; idx < rs.length; idx += 1) {
-      result = concat(result, of(F, rs[idx].x));
     }
-    return result;
+    var rs = reduce(reducer, [], foldable);
+    if (rs.length === 0) {
+      return foldable;
+    }
+    var lteFn =
+      typeof rs[0].prop === 'string' || typeof rs[0].prop === 'number' ?
+        function(a, b) { return a <= b; } :
+        lte;
+    rs.sort(function(a, b) {
+      return lteFn(a.prop, b.prop) ?
+        lteFn(b.prop, a.prop) ? a.idx - b.idx : -1 :
+        1;
+    });
+    if (foldableIsArray) {
+      // If the original foldable was an array we don't need to convert the
+      // array back to a foldable.
+      for (var j = 0; j < rs.length; j += 1) {
+        rs[j] = rs[j].x;
+      }
+      return rs;
+    } else {
+      var result = empty(foldable.constructor);
+      var of = foldable.constructor['fantasy-land/of'];
+      for (var i = 0; i < rs.length; i += 1) {
+        result = result['fantasy-land/concat'](of(rs[i].x));
```

Comment:
Is this safe? Are we sure that Array is the only built-in type that can support sorting?